include/rc_export.h

```diff
@@ -0,0 +1,69 @@
+#ifndef RC_EXPORT_H
+#define RC_EXPORT_H
+
+/* These macros control how callbacks and public functions are defined */
+
+/* RC_SHARED should be defined when building rcheevos as a shared library (e.g. dll/dylib/so) */
+/* RC_STATIC should be defined when building rcheevos as a static library */
+/* BUILDING_RC should be defined when building rcheevos itself, and should not be defined for external code */
+
+/* TODO: BUILDING_RC currently only has an effect for RC_SHARED (this is abused for test code), perhaps when it is undefined internal definitions should be hidden? */
+
+#if !defined(RC_SHARED) && !defined(RC_STATIC)
+  #error RC_SHARED or RC_STATIC must be defined
+#endif
+
+#if defined(RC_SHARED) && defined(RC_STATIC)
+  #error RC_SHARED and RC_STATIC are mutually exclusive
+#endif
+
+/* RC_C_LINKAGE should be used for callbacks, to enforce the C calling convention */
+/* RC_C_LINKAGE should be placed before the return type, or before typedef if present */
+/* RC_C_LINKAGE void (*rc_callback)(void) */
+/* RC_C_LINKAGE typedef void (*rc_callback_t)(void) */
+
+#ifdef __cplusplus
+  #define RC_C_LINKAGE extern "C"
+#else
+  #define RC_C_LINKAGE
+#endif
```

Comment:
This is also very verbose and prone to being forgotten. How is this better than doing the `extern "C" {` for the whole file?

Reply:
I think my intent was to make the extern C usage explicit for function pointers rather than implicit, although I guess either way I did that anyways for RC_EXPORT, so it's probably not that much better in the first place (also extern "C" for function pointers is more academically correct if anything, in practice nearly every C++ implementation does not actually differ the calling convention depending on C / C++ linkage.

Linkage is the actual term used for extern "C" / extern "C++" anyways, and it makes sense given how it actually works. All extern "C" really does in practice is disable the name mangling (and possibly error for code that requires name mangling, e.g. overloads, templates), so symbols can be linked against C code, as it follows the same linkage as C code. It technically also enforces a C calling convention rather than a C++ calling convention, but for nearly every implementation these are the same.

Reply:
I suppose I could still use this idea to define RC_BEGIN_C_HEADER and RC_END_C_HEADER macros (or maybe some other name), which mght be slightly better than the contant #ifdef __cplusplus checks.